pkg/elb/cleanup.go

```diff
@@ -0,0 +1,314 @@
+	loadBalancers := map[string]loadBalancer{}
+	for _, s := range services.Items {
+		lb, err := getServiceLoadBalancer(ctx, ec2API, elbAPI, clusterConfig.Metadata.Name, &s)
+		if err != nil {
+			return err
+		}
+		if lb == nil {
+			continue
+		}
+		logger.Debug("tracking deletion of Load Balancer %s of kind %d with security groups %v",
+			lb.name, lb.kind, convertStringSetToSlice(lb.ownedSecurityGroupIDs))
+		loadBalancers[lb.name] = *lb
+		logger.Debug("deleting 'type: LoadBalancer' service %s/%s", s.Namespace, s.Name)
+		if err := kubernetesCS.CoreV1().Services(s.Namespace).Delete(s.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+
+	// Wait for all the load balancers backing the LoadBalancer services to disappear
+	pollInterval := 2 * time.Second
+	for ; time.Now().Before(deadline) && len(loadBalancers) > 0; time.Sleep(pollInterval) {
+		for name, lb := range loadBalancers {
+			exists, err := loadBalancerExists(ctx, ec2API, elbAPI, elbv2API, lb)
+			if err != nil {
+				logger.Warning("error when checking existence of load balancer %s: %s", lb.name, err)
+			}
+			if exists {
+				continue
+			}
+			logger.Debug("load balancer %s and its security groups were deleted by the cloud provider", name)
+			// The load balancer and its security groups have been deleted
+			delete(loadBalancers, name)
+		}
+	}
+
+	if len(loadBalancers) > 0 {
+		return fmt.Errorf("deadline surpased waiting for load balancers to be deleted")
+	}
+	logger.Debug("deleting Load Balancer Security Group orphans")
+	// Orphan security-group deletion is needed due to https://github.com/kubernetes/kubernetes/issues/79994
+	// and because we could have started the service deletion when a service didn't finish its creation
+	return deleteOrphanLoadBalancerSecurityGroups(ctx, ec2API, clusterConfig)
+}
+
+func getServiceLoadBalancer(ctx context.Context, ec2API ec2iface.EC2API, elbAPI elbiface.ELBAPI, clusterName string,
+	service *corev1.Service) (*loadBalancer, error) {
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return nil, nil
+	}
+	name := cloudprovider.DefaultLoadBalancerName(service)
+	kind := getLoadBalancerKind(service)
+	ctx, cleanup := context.WithTimeout(context.Background(), 30*time.Second)
+	securityGroupIDs, err := getSecurityGroupsOwnedByLoadBalancer(ctx, ec2API, elbAPI, clusterName, name, kind)
+	cleanup()
+	if err != nil {
+		return nil, err
+	}
+	lb := loadBalancer{
+		name:                  name,
+		kind:                  kind,
+		ownedSecurityGroupIDs: securityGroupIDs,
+	}
+	return &lb, nil
+}
+
+func convertStringSetToSlice(set map[string]struct{}) []string {
+	result := make([]string, 0, len(set))
+	for value := range set {
+		result = append(result, value)
+	}
+	return result
+}
+
+func deleteOrphanLoadBalancerSecurityGroups(ctx context.Context, ec2API ec2iface.EC2API, clusterConfig *api.ClusterConfig) error {
+	clusterName := clusterConfig.Metadata.Name
+	describeRequest := &ec2.DescribeSecurityGroupsInput{
```

Comment:
Our current convention is just to call these `input`, and whatever we get back `output`, and avoid having more than one API call in one function.

Reply:
Just to say that there is a convention, don't need to worry about it right now.

Reply:
See #1035.